src/cockpit/samples.py

```diff
@@ -301,6 +301,60 @@ def sample(self, samples: Samples) -> None:
                     samples['cgroup.cpu.usage'][cgroup] = usage_nsec / 1000000
 
 
+class CGroupDiskIO(Sampler):
+    descriptions = [
+        SampleDescription('disk.cgroup.read', 'bytes', 'counter', True),
+        SampleDescription('disk.cgroup.written', 'bytes', 'counter', True),
+    ]
+
+    @staticmethod
+    def get_cgroup_name(fd: int) -> Optional[str]:
+        cgroup_name = ""
+        try:
+            with Handle.open('cgroup', os.O_RDONLY, dir_fd=fd) as cgroup_fd:
+                cgroup_name = os.read(cgroup_fd, 1024).decode().strip()
+        except FileNotFoundError:
+            return None
+
+        # Skip leadin ::0/
+        return cgroup_name[4:]
+
+    @staticmethod
+    def get_proc_io(fd: int) -> Optional[List[int]]:
+        data = ""
+        try:
+            with Handle.open('io', os.O_RDONLY, dir_fd=fd) as io_fd:
+                data = os.read(io_fd, 1024).decode()
+        except (FileNotFoundError, PermissionError):
+            return None
+
+        values = []
+        for key in ['read_bytes:', 'write_bytes:']:
+            start = data.index(key) + len(key)
+            end = data.index('\n', start)
+            values.append(int(data[start:end]))
```

Comment:
Although I think the general strategy here is fairly OK (albeit possibly buggy in presence of keys with names like `other_read_bytes:` or something) I really don't like the list.  I wonder if maybe you could do this more reliably with two regexps instead of all of this slicing and dicing.  Maybe even one, if you use named groups.

Something like this works, for example:

```python
>>> int(re.match(rb'\bread: (\d+)', b'read: 5').group(1))
5
```

...and you can also save the `.decode()` by working on bytes.

I'm not sure that this is the best of all possible solutions, but I definitely dislike it a lot less than the list.  Maybe you can think of something better.